### PR TITLE
Set up the framework of type casting for ORC reading

### PR DESCRIPTION
**Pull request by firestarman**
Contributes to https://github.com/NVIDIA/spark-rapids/issues/5895, and fixes https://github.com/NVIDIA/spark-rapids/issues/5948.

Now this PR supports only type casting between integral types. Will add more in the future, tracked in #5895.

Unfornately we can not fallback to CPU for unsupported casting, because we do not know the file schema at that time. `dataSchema` is not reliable because files may have different schemas.

What's mainly changed.
- Updated `SchemaUtils.evolveSchemaIfNeededAndClose` to try to call a caller-provided cast function when the column type is not equal to the required one.
- Added two methods `canCast` and `castColumnTo` to do the actual type casting things for ORC reading. `castColumnTo` will be executed as a part of schema evolution in `evolveSchemaIfNeededAndClose`.
- Updated `checkSchemaCompatibility` to also pass the check if `canCast` returns true for read schema being different from file schema.   
- Replaced the `requestedColumnIds` with the Spark one, we don't need to implement it in plugin because the Spark one is public. Also added shims for this method because its signature has been changed from 330.
- Removed `resolveTableSchema` which is a duplicate of `buildReaderSchema`.
- Rewrote the computation of `fileIncluded`. The CPU evlotuion's `fileIncluded` can not be used here, because it has different type casting support than GPU.
- Removed the column re-order logic in `columnRemap` because we will also do it in `evolveSchemaIfNeededAndClose`.
- Created `decodeToBatch` and `decodeToTable` in a parent trait to handle the common things of cuDF decoding for 3 kinds of readers, this simplifies the code a lot.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 1]
```diff
@@ -338,8 +428,13 @@ case class OrcPartitionReaderContext(
     requestedMapping: Option[Array[Int]])
 
 /** Collections of some common functions for ORC */
-trait OrcCommonFunctions extends OrcCodecWritingHelper {
-  def execMetrics: Map[String, GpuMetric]
+trait OrcCommonFunctions extends OrcCodecWritingHelper { self: FilePartitionReaderBase =>
+  private val orcFormat = Some("avro")
```
Comment (wbo4958) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:432]: why there is avro?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 2]
```diff
@@ -153,6 +152,98 @@ object GpuOrcScan {
       meta.willNotWorkOnGpu("mergeSchema and schema evolution is not supported yet")
     }
   }
+
+  private lazy val numericLevels = Map(
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:156]: nit: looks like we only care about the relative ordering of levels. So we could just do a 
```scala
private lazy val numericLevels = Seq(
   DType.DTypeEnum.BOOL8, 
   DType.DTypeEnum.INT8,
   ...
 ).zipWithIndex.toMap 
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:431]: nice catch

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala (hunk 2 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:156]: nice suggestion

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
The failure is not related.
`[2022-07-07T10:45:14.683Z] FAILED ../../src/main/python/map_test.py::test_get_map_value_string_col_keys_ansi_fail[Map(String(not_null),String)]`

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
@revans2 Hi, Could you help review this ?

**Review by revans2 — COMMENTED** [review]
I have done a quick pass through this and I haven't seen anything off, but I am no expert on the ORC parser. I would feel better if @jlowe and/or @tgravescs also take a look at this. I am going to keep looking at this, but it is going to take some time for me to feel comfortable approving this.

**Review by jlowe — COMMENTED** [review]
Looks good overall, just some cleanup nits.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 3]
```diff
@@ -1217,104 +1149,95 @@ private case class GpuOrcFileFilterHandler(
...
-                  val (newChild, childInclude) =
-                    checkSchemaCompatibility(fileReadType, readType, isCaseAware, include)
-                  prunedReadSchema.addField(readField, newChild)
-                  prunedInclude ++= childInclude
+        isCaseAware: Boolean): (TypeDescription, Array[Boolean]) = {
+      // all default to false
+      val fileIncluded = new Array[Boolean](fileSchema.getMaximumId + 1)
+      val isForcePos = OrcShims.forcePositionalEvolution(conf)
+      // An inner function
+      lazy val checkTypeCompatibility: (TypeDescription, TypeDescription) => TypeDescription =
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1174]: `lazy` is required to be an inner function `val`, but no longer needed after moving it out being a seperate function.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 4]
```diff
@@ -467,15 +549,79 @@ trait OrcCommonFunctions extends OrcCodecWritingHelper {
     buffer.toArray
   }
 
+  /**
+   * Read the host data to GPU for ORC decoding, and return it as a cuDF Table.
+   * The input host buffer should contain valid data, otherwise the behavior is
+   * undefined.
+   * 'splits' is used only for debugging.
+   */
+  protected def sendToGpuUnchecked(
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:558]: Nice suggestion, updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 5]
```diff
@@ -153,6 +152,98 @@ object GpuOrcScan {
...
+    //    val b = (toDt)a
+    //    val c = (fromDt)b
+    //    c != a
+    withResource(col.castTo(toType)) { casted =>
+      val overflowFlags = withResource(casted.castTo(col.getType)) { backed =>
+        col.equalTo(backed)
+      }
+      // Replace values that cause overflow with nulls, same with CPU ORC.
+      // Do this no matter whether there is any real overflow to avoid an `if` check
+      // which will take some addition workload.
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:184]: Sure, removed.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
I am going to remove `databricks` in the title to avoid duplicate runs on databricks, since we have already verified it in previous runs.

**Comment by firestarman** [comment]
The CI failure is not related.

**Review by jlowe — APPROVED** [review]

**Comment by firestarman** [comment]
Hi @revans2, do you have any concern ?

**Comment by firestarman** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by firestarman** [comment]
I am going to merge this, it is only the first step for type casting. 
If any concern, I can address them in new PRs.

**Pull request merged**